### DISCUSSION TITLE
Actions called by listeners don't trigger loading states when targeted

### DISCUSSION THREAD

**Discussion by isaackearl** [livewire/livewire#3080] [7. Bugs (archived)]
### Description

Short version: there is no way to show targeted loading states on things that happen because of livewire events... which makes working with multiple components on a page that communicate via emitting etc a bad experience for the user.

When I have a method aka an "Action" that gets triggered by a listener that I have registered on the component, it does not trigger the loading states when I have a target set for that action.  I have circumstances where I really need to show a component is loading when the action is firing... But right now since .target modifier doesn't work when the action is triggered from an event listener.. I have no way of showing the loading.  I can remove the target modifier and get un-targeted loading, but the problem with that is then it shows the loading symbol during every request which totally doesn't work for me either.  I've event tried targeting a public property that gets changed and that doesn't work either.

### Exact steps to reproduce

Here is a simple laravel playground I've setup so you can see that it doesn't work.  If I remove the wire:target bit then it works normally, however I find myself in situations quite often where I only want it to show loading stuff during the processing of an event, and NOT when i'm doing normal stuff like modifying a form.
https://laravelplayground.com/#/snippets/ada8bdf1-578e-4570-b8ee-d32581639ce4

### Stripped-down, copy-pastable code snippets

see the very simple laravel playground code above

**Context**
- Livewire version: 2.2.9
- Laravel version: 8.7.1
- Browser: Chrome (any browser)

**Comment by FarhanShares** [comment]
I think it's relatable to #183 

I've also added a  [comment](https://github.com/livewire/livewire/issues/183#issuecomment-703394831), perhaps a use case of it.
Badly need that, it's a bad impact on UX.

**Comment by isaackearl** [comment]
@farhanianz I think it is similar but not quite the same.  I'm not trying to target events for loading, but rather the actions that are called by listeners on the php side.  Actions are already something you should be able to target for loading, and so it seems intuitive that that would work.

**Comment by isaackearl** [comment]
For now i have a semi-workaround for this problem: 

Let's say I have a listener on the php side that triggers an action "handleSelection".  if I setup a loading target such as wire:target="handleSelection", then it never works...

On the other hand, if I listen for the event in javascript like so: (i remove the listener on the php side)
```
        window.Livewire.on("selected", (item) => {
          this.$wire.handleSelection(item);
        });
```
... then when I call the action this way the loading target works.  So this is sort of a "hack" but it will do for now I guess.

**Comment by isaackearl** [comment]
Cancel that.  My workaround actually causes another problem, because each time the component is shown on the page, it adds another event listener, so eventually the method gets called multiple times instead of once.  Is there a way to unregister event listeners?  Or to only register the event listener if it doesn't already exist?

**Reply by BrianHannay** [reply]
Does this work for you?
```
document.addEventListener('livewire:init', function ()
{
  window.Livewire.on("selected", (item) => {
    this.$wire.handleSelection(item);
  });
}
```

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by isaackearl** [comment]
REOPEN

**Comment by isaackearl** [comment]
I re-opened because I believe not only that this is a bug, but that it also goes against the goals and spirit of livewire.  The fact that you can't have loading states triggered by the actions that are called because of a listener makes it very difficult to use multiple livewire components that interact on a page.  It forces you to make large livewire components only (unless you don't care about showing proper loading states).  I will eventually try and help fix this bust as I said on other issues, I'm currently working way more then I can manage.

**Comment by emkcloud** [comment]
the function would be useful, I also have cases where I launch for example a cashout event from interactions external to the wire component and I cannot activate the wire:loading correctly, it only works without wire:target but at this point it is activated even when I change the fields of the form on which I would not like it to be activated ...

`<div click="$emit('cashout')"></div>`
```
protected $listeners = [
   'cashout' => 'save',
];
```
```
<div wire:loading wire:target="save">
     <x-content.area.loading/>
</div>
```

**Comment by ahinkle** [comment]
also keep running into this bug unfortunately. was anyone able to find a workaround?

**Comment by IP-Developer** [comment]
+1 for this

**Comment by renamate** [comment]
any idea? 
here are a discussion linked to this challenge! [Link](https://www.reddit.com/r/laravel/comments/kb0o97/livewire_or_inertia_for_admin_panel/)

**Comment by quantrung9** [comment]
Is there any workaround for this one?

**Comment by sunil19822701** [comment]
+1

**Comment by Alloco225** [comment]
+1

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by isaackearl** [comment]
@nuernbergerA actually the provided example does NOT work.  in the provided example it is supposed to show "Loading..." while the action is being called.

**Comment by emkcloud** [comment]
REOPEN

**Comment by ikechukwukalu** [comment]
@isaackearl Your hack solved it for me. Just a very slight modification.

```
Livewire.on("event_name", (item) => {
    @this.function_name(item);
});
```

**Comment by mikebronner** [comment]
I'm looking for a solution for this problem as well. I could not get the mentioned work-arounds to trigger the loading indicator.

**Comment by ryatkins** [comment]
Below is an Alpine JS option using x-show to display an animated spinner.   

You can target a method or an event.

```
<div x-data="data()">

   <div x-show="rowLoading" x-cloak>
    <svg  class="animate-spin h-5 w-5 text-red-700" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24">
        <circle class="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" stroke-width="4"></circle>
        <path class="opacity-75" fill="currentColor" d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"></path>
    </svg>
   </div>

</div>

```


```
  function data () {
        return {

            rowLoading: false,

            init() {
            

                Livewire.hook('message.sent', (message, component) => {

                    let method = message.updateQueue.filter((q => {

                        if (q.method == 'saveBOM') {
                            this.rowLoading = true
                            return true
                        }

                        if (q.payload.event == 'refreshWithNewBomRow') {
                            this.rowLoading = true
                            return true
                        }
                    }))
                });

                Livewire.hook('element.updated', (message, component) => {
                    this.rowLoading = false;                   

                });
            }
}
```

**Comment by ryatkins** [comment]
Alternatively, a call to Livewire method using `$wire` returns a promise and can be used with x-show:

```
<div x-data="data()">
   <div x-show="rowLoading" x-cloak>
           <svg ...
  </div>

  <button @click="saveData()">Save</button>
</div>
```

```
function data () {
        return {

            rowLoading: false,

            saveData() {
                        this.rowLoading = true

                        this.$wire.save().then( () => this.rowLoading = false )
            }

```

**Comment by booni3** [comment]
Are there any better workarounds to this? My use case in pseudo:

### Component 1 - menu
```php
<button wire:click="$emit('showGallery', 'all')">Show All - Loading Icon does not show/button>
<button wire:click="$emit('showGallery', 'group1')">Show Group 1 - Loading Icon does not show</button>
```

### Component 2 - gallery
```php
protected $listeners = ['showGallery' => 'getMedia'];

public function getMedia($filter) {
   $this->media = Media::query()->where('filter', $filter)->get()
}
```

```blade
<div wire:loading wire:target="getMedia">loading</div>
<button wire:click="getMedia('all')">Show All Media - Loading icon shows</button>
@dump($media)
```

When calling getMedia from within the component, the loading icon shows. When calling it via a listener, it does not.

The gallery is a self-contained component, and I would expect the loading icon to show any time when getMedia is being called, now matter how it is being called.

It seems like odd behaviour that it does not do this by default. Am I missing something?

**Comment by Jiysea** [comment]
For Livewire v3, since I used a JS library for dates and needs an event to trigger changes, just call the function with the `$wire` object just as @ryatkins suggested:

```javascript
const startDate = document.getElementById('start-date');
startDate.addEventListener('changeDate', function(event) {
            $wire.setStartDate(startDate.value);
        });
```

or an alternative approach using AlpineJS event listener:

```css
<input type="text" 
       id="start-date"
       @change-date.camel="$wire.setStartDate($el.value);">
```

and then just add `setStartDate` to the `wire:target` in my case
